### ISSUE TITLE
BKD-1 — Kernel-backend Protocol, registry, and the plain-torch reference backend

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1560]
**Depends on:** INF-1 (#1550), CORE-4 (#1558) · **Blocks:** ARCH-2 (#1561), BKD-2 (#1566), BKD-3 (#1567), BKD-4 (#1568), FM-00 (#1565)

**Context:** Today the equivariant ops are dispatched by `mace/modules/wrapper_ops.py`, and the backend choice **leaks into the model**: interaction blocks branch on `hasattr(self, "conv_fusion")` at six sites (`mace/modules/blocks.py:918,1021,1140,1264,1393,1577`, the attribute set at `:675-678`) and `wrapper_ops.py` still monkeypatches `.forward` with `types.MethodType` at three (`:130` `with_scatter_sum`, `:244` `with_oeq_conv_fusion`, `:266` `with_oeq_scatter_sum`) — the cueq fusion path was since promoted to a real `nn.Module`, `CueqConvFusionWrapper` (`:134`), which is the shape v1 generalizes, not the monkeypatch. Selection itself is five `__new__` if/elif branches (`Linear:82`, `TensorProduct:273`, `FullyConnectedTensorProduct:381`, `SymmetricContractionWrapper:418`, `TransposeIrrepsLayoutWrapper:459`). There are five weight-conversion CLIs because each backend stores weights differently. This ticket builds the clean dispatch layer that replaces all of that: a small closed op contract, an entry-point registry, and a mandatory **plain-torch reference backend** — compile-first, differentiable to second order, **no e3nn**. Legacy is read in place as the behavioural oracle; reproduce its numbers, never its structure.

**Interface & constraints:**

1. **The op set is closed** — exactly the ops that differ between backends:

| Op | Dispatched? | Weights | Note |
|---|---|---|---|
| `channelwise_tp_conv` | yes | external (radial MLP) | **always node-level**: returns `[n_nodes, …]`; whether the scatter is fused is backend-internal (deletes the `conv_fusion` branch) |
| `symmetric_contraction` | yes | internal, per-element, canonical flat `[Z, A, mul]` | hot path #2 |
| `linear` | yes | internal; **bias is first-class** | |
| `fully_connected_tp` (skip_tp) | yes | internal | |
| `segment_reduce` | yes | — | no irreps semantics; reused for energy/pair-repulsion reductions |
| `spherical_harmonics`, `radial_basis` | reference-only | — | closed-form, cheap; a backend *may* override (evaluate `sphericart` later) but is not required to |
| `interaction_layer` (span) | optional | via `to_canonical` | optional factory fusing a whole layer; `NotImplemented` → op-by-op fallback |

2. **The Protocol** (in `mace_core.kernels.protocol`; imports no torch/jax; dtypes are names `Literal["float64",…]`, never `torch.dtype`):

```python
class KernelBackend(Protocol):                 # generic over TensorT
    name: str
    def capabilities(self) -> BackendCapabilities: ...
    # factories: descriptor -> op instance, resolved ONCE at model build time
    def make_linear(self, d: LinearDescriptor) -> LinearOp: ...
    def make_channelwise_tp_conv(self, d: ChannelwiseTPConvDescriptor) -> ChannelwiseTPConv: ...
    def make_symmetric_contraction(self, d: SymmetricContractionDescriptor) -> SymmetricContractionOp: ...
    def make_fully_connected_tp(self, d: FullyConnectedTPDescriptor) -> FullyConnectedTPOp: ...
    def make_segment_reduce(self, d: SegmentReduceDescriptor) -> SegmentReduceOp: ...
    def make_spherical_harmonics(self, d) -> SphericalHarmonicsOp: ...   # reference-only, may decline
    def make_radial_basis(self, d) -> RadialBasisOp: ...                 # reference-only, may decline
    def make_interaction_layer(self, ds) -> "InteractionLayerOp | NotImplemented": ...   # optional span
```

Internal-weight ops (`linear`, `symmetric_contraction`, `fully_connected_tp`) also implement `to_canonical()` / `load_canonical()`, applied **only** at the checkpoint save/load boundary, never in `forward`. They also implement **`initialize_weights(seed)`**, which draws a fresh set in the canonical layout. The draw is not optional: an op that allocates zeros and waits to be loaded is correct for every model that comes from a checkpoint and builds a model that cannot be trained, since zero-valued tensors multiply to zero and so does the gradient. A run started from scratch then finishes and reports a model that never moved, and no test that loads a checkpoint or overwrites the parameters first can see it. The seed is chosen per op by the caller, from the run's seed and the op's path, so two ops of the same shape do not receive the same weights and the draw does not depend on the global generator.

3. **Compile-safety (hard requirement).** Each dispatched hot op is a `torch.library.custom_op` with a `register_fake` (meta) impl and a `register_autograd` with a **differentiable backward** (double-backward / force training must be correct). Consequences: `torch.compile(fullgraph=True)` sees one opaque node per op (no graph break); `num_nodes` enters as a symbolic dimension (`int`→`SymInt`), never read from a device tensor or `edge_index.max()`; `register_fake` builds outputs at the symbolic size (no recompile across batch sizes); **no data-dependent host reads inside op bodies** (breaks CUDA-graph capture under `reduce-overhead`).

4. **Capabilities.** `capabilities()` returns coarse fields (`ops`, `devices`, `dtypes`, `max_lmax`, `layouts`, `bases`, `supports_double_backward`) as a cheap first filter, plus the **authoritative `supports(descriptor) -> bool`** over the full op descriptor. **Missing double-backward while force/stress training → hard reject at build time** (loud error, never silently wrong forces). The contract carries a `KERNEL_SPEC_VERSION`.

5. **Canonical weights.** One canonical layout means one checkpoint loads into any backend: `mul_ir`, the reduced-Clebsch-Gordan-basis weight order MACE pins (CORE-4 (#1558), computed from first principles — no e3nn/cueq), flat `[Z, A, mul]` for the symmetric contraction, `weight_numel` a descriptor property. The reference holds canonical directly (`to_canonical`/`load_canonical` are identity/views).

   **The layout carries its normalization folded into the weight**, where the frozen tree keeps a standard normal and applies a per-path factor inside the `e3nn` operation. That makes the factor part of the format rather than of an implementation, so the scales a fresh draw uses are stated with the layout and every backend draws at the same size: a linear divides by the square root of the total input multiplicity feeding the output irrep, the skip connection's tensor product against the element attributes by the square root of the product of the two input multiplicities (`e3nn`'s per-instruction factor and the scalar coupling cancel the output dimension between them), and the symmetric contraction by nothing. A bias starts at zero, since it is an offset on the energy.

6. **The reference backend** (`mace_torch.backends.reference`): mandatory, CPU-capable, native plain torch — **no e3nn**. Symmetric contraction via a Horner cascade over the reduced basis; spherical harmonics + radial bases native closed-form. It is the correctness oracle every other backend is checked against.

7. **Registry.** Backends register via entry points, per framework: group `mace.kernel_backends.torch`. Discovery **records** import failures (a missing CUDA lib is not fatal); resolution raises. Dispatch resolves **once at model build time**, frozen into the module tree — nothing resolves in `forward` (no dtype/device/isinstance checks on the hot path).

8. **The Clebsch-Gordan basis is model state, resolved once and recorded.** Develop decides it in three places that disagree, which is exactly why v1 must make it one field: the CLI default is `--use_reduced_cg=False` (`mace/tools/arg_parser.py:240-244`), the model-class default is `use_reduced_cg=True` (`mace/modules/models.py:67`), the converters default to `True` when the key is absent from a checkpoint config (`mace/cli/convert_e3nn_cueq.py:241`, `convert_cueq_e3nn.py:246`, `convert_e3nn_hybrid.py:51`) — and `SymmetricContractionWrapper.__new__` silently forces it back to `False` on a host without `cuequivariance` (`mace/modules/wrapper_ops.py:428`, `use_reduced_cg = use_reduced_cg and CUET_AVAILABLE`). Orthogonally, the environment variable `MACE_USE_CUEQ_CG` (`mace/tools/cg.py:23`) switches the basis used to build `U_matrix_*`. `mace/tools/scripts_utils.py:293-295` records only whatever the built model ended up with. In v1 the basis is one resolved-config field, a backend **declares** which bases it supports and the build fails loudly on a mismatch; nothing infers it from availability and no environment variable changes numerics.

**Task:**
1. `mace_core.kernels`: descriptors, `BackendCapabilities`, the `KernelBackend` Protocol, the entry-point registry (`get_backend(name, framework)`), `KERNEL_SPEC_VERSION`, the canonical-layout spec — all torch-free.
2. `mace_torch.kernels`: the 5 dispatched ops as `custom_op` + `register_fake` + `register_autograd` (double-backward); `num_nodes` as `SymInt`.
3. `mace_torch.backends.reference`: the plain-torch reference backend implementing every factory (Horner symmetric contraction over CORE-4 (#1558)'s reduced basis; native spherical harmonics/radial; first-class bias in `linear`), including `initialize_weights` on the three weighted ops, plus the walk over a built module tree that gives each one its per-op seed.
4. Node-level `channelwise_tp_conv` (returns `[n_nodes,…]`); no `conv_fusion` branch, no `.forward` monkeypatch anywhere.
5. Register the reference backend under `mace.kernel_backends.torch`; a dummy backend registered via an entry point is discovered.

**Out of scope:** deleting the five legacy `convert_*` CLIs (RET-3 (#1599)); the cueq/oeq backends (BKD-3 (#1567)); the compiled padded path + PrecisionConfig (BKD-2 (#1566)).

**Acceptance criteria:**
- [ ] `gradcheck` **and** `gradgradcheck` pass at fp64 for `symmetric_contraction` and `channelwise_tp_conv`, and for the full pos→spherical-harmonics→tensor-product→scatter force path w.r.t. positions, weights and features.
- [ ] `torch.compile(fullgraph=True)` produces no graph break across ≥5 batch sizes with a single frame compile (no per-size recompilation).
- [ ] A backend declaring `supports_double_backward=False` is hard-rejected at build time for force/stress training (loud error); usable for inference.
- [ ] A checkpoint written by the reference loads and reproduces its outputs through the registry (round trip) at fp64 tolerance.
- [ ] A freshly built model has no all-zero weighted op: every op that holds weights draws a set, two ops of the same shape in one model draw different ones, the same seed rebuilds the same model, and the draw leaves the global generator untouched. The drawn scales are the ones the canonical layout states.
- [ ] `mace_core.kernels` imports neither torch nor jax (asserted); the reference backend uses no e3nn.
- [ ] Reference outputs match the frozen legacy `mace/` on the tiny anchors at the fp64 tolerance row.
- [ ] The resolved Clebsch-Gordan basis is a single recorded config field; a build whose backend does not support the recorded basis fails loudly, and no code path reads `MACE_USE_CUEQ_CG` or infers the basis from `CUET_AVAILABLE`.

**Inventory gaps assigned here:**
- [ ] The Clebsch-Gordan basis source is a build-time decision recorded in the resolved config — covering `--use_reduced_cg` and the `MACE_USE_CUEQ_CG` environment variable (`mace/tools/cg.py:23`), neither of which is tested and both of which change numerics silently today.

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/backends/test_reference_conformance.py -v   # gradcheck/gradgradcheck, compile-no-break
python -m pytest packages/mace-torch/tests/test_mace_torch_weight_initialization.py -v # a built model can be trained at all
python -m pytest packages/mace-core/tests/test_kernels_registry.py -v                  # discovery, no torch import
python -m pytest tests/parity -m "not slow" -k reference                               # vs frozen legacy
```
**Review focus:** the double-backward autograd (backward-of-backward) of `symmetric_contraction` and `channelwise_tp_conv`, and that nothing resolves in `forward` — numerics-critical, needs a physics reviewer.

---
